src/Containers/Microsoft.NET.Build.Containers/ImageBuilder.cs

```diff
@@ -87,6 +87,11 @@ internal void AddLayer(Layer l)
         _baseImageConfig.AddLayer(l);
     }
 
+    internal void AddBaseImageDigestLabel()
+    {
+        AddLabel("org.opencontainers.image.base.digest", _manifest.GetDigest());
```

Comment:
@baronfel I wonder if we're mixing up annotations and labels in this PR?

Ref https://github.com/opencontainers/image-spec/blob/main/annotations.md.

Reply:
Most tools I've seen support both. An example of this is snyk: https://snyk.io/blog/how-and-when-to-use-docker-labels-oci-container-annotations/

Annotations are definitely the way to go when creating multi-arch image indexes/manifest lists, but we're not doing that yet. And most tools do seem to understand Labels, especially for single-platform image definitions. So this is a bit of a gap-filler, and when we do add multi-platform index support we'll need to decide which labels/annotations end up in which spot.

Reply:
> Most tools I've seen support both.

fyi, I tried with `podman` by making it export to the Docker format, but it did not include the label.
For the oci format, it does include the annotation.

> _manifest.GetDigest()

I assume all clients are supposed to come to the same digest value for the same `FROM` image?

Does the value computed by the SDK equal what `docker`/`podman` outputs as `org.opencontainers.image.base.digest`?

Reply:
> fyi, I tried with podman by making it export to the Docker format, but it did not include the label.
> For the oci format, it does include the annotation.

I'd love to dig into this more - do you have a quick sample repro that I could play with?

> I assume all clients are supposed to come to the same digest value for the same FROM image?
> Does the value computed by the SDK equal what docker/podman outputs as org.opencontainers.image.base.digest?

This is a good sanity check - I'd expect it to be so since we're basing almost everything else off of this digest calculation already but it's good to verify.

Reply:
> This is a good sanity check - I'd expect it to be so since we're basing almost everything else off of this digest calculation already but it's good to verify.

Nope! I pulled the latest aspnet base image:

```terminal
> docker inspect mcr.microsoft.com/dotnet/aspnet:8.0 | jq -r '.[0].Id'
sha256:789045ecae51d62d07877994d567eff4442b7bbd4121867898ee7bf00b7241ea
```

then did a publish of an asp.net app with this change:

```terminal
> dotnet publish -t:PublishContainer -r linux-arm64 -bl
  Determining projects to restore...
  Restored E:\Code\sdk-container-demo\src\sdk-container-demo\sdk-container-demo.csproj (in 739 ms).
  sdk-container-demo -> E:\Code\sdk-container-demo\artifacts\bin\sdk-container-demo\release_linux-arm64\sdk-container-demo.dll
  sdk-container-demo -> E:\Code\sdk-container-demo\artifacts\publish\sdk-container-demo\release_linux-arm64\
  Building image 'sdk-container-demo' with tags 'latest,v1.1' on top of base image 'mcr.microsoft.com/dotnet/aspnet:8.0'.
  Pushed image 'sdk-container-demo:latest, v1.1' to local registry via 'docker'.
```

then inspected the built image's labels:

```terminal
>docker inspect sdk-container-demo | jq -r '.[0].Config.Labels'
{
  "org.opencontainers.artifact.created": "2024-03-05T16:03:08.7216269Z",
  "org.opencontainers.artifact.description": "A project that demonstrates publishing to various container registries using just\r\n      the .NET SDK",
  "org.opencontainers.image.authors": "Chet Husk",
  "org.opencontainers.image.base.digest": "sha256:4af8a040f44fc1b45f9d9f31762450335a3d98d7da34255505ffa72d99705732",
  "org.opencontainers.image.base.name": "mcr.microsoft.com/dotnet/aspnet:8.0",
  "org.opencontainers.image.created": "2024-03-05T16:03:08.7216269Z",
  "org.opencontainers.image.description": "A project that demonstrates publishing to various container registries using just\r\n      the .NET SDK",
  "org.opencontainers.image.documentation": "https://github.com/baronfel/sdk-container-demo",
  "org.opencontainers.image.licenses": "MIT",
  "org.opencontainers.image.revision": "a378325876ea56873be5069da0255aac6ecd1f4d",
  "org.opencontainers.image.source": "https://github.com/baronfel/sdk-container-demo.git",
  "org.opencontainers.image.title": ".NET SDK 8 Container Demo",
  "org.opencontainers.image.url": "https://github.com/baronfel/sdk-container-demo",
  "org.opencontainers.image.version": "1.0.0"
}
```

There's a mismatch for sure.

Reply:
Did some digging/added some asserts and found the following:
* the digest calculation we use is correct (`Debug.Assert(DigestUtils.GetDigest(json) == matchingManifest.digest)` passes)
* the digest we compute for a `ManifestV2` structure _from_ the raw json is incorrect (`Debug.Assert(manifest.GetDigest() == matchingManifest.digest` fails).

e.g.

Given this manifest json:

```json
{
   "schemaVersion": 2,
   "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
   "config": {
      "mediaType": "application/vnd.docker.container.image.v1+json",
      "size": 2717,
      "digest": "sha256:01b704c914bb479f6361031838854e96bbba53bceef4f6cfd4fd67ef314f51b0"
   },
   "layers": [
      {
         "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
         "size": 29156363,
         "digest": "sha256:f546e941f15b76df3d982d56985432b05bc065e3923fb35be25a4d33d5c0f911"
      },
      {
         "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
         "size": 18480791,
         "digest": "sha256:243835fecb16018c133dd8834d6fac258d646a43af19edfb8bbe992e9c6e8979"
      },
      {
         "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
         "size": 3279,
         "digest": "sha256:ba86dd0a850b71f02df84db8858f96fc435e37b30cb594ac638878aff2981b20"
      },
      {
         "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
         "size": 30807516,
         "digest": "sha256:6bcd598847e3669a1dacd9f4dd1dd81b1902c8f877b5f7721804a27207cd8a1b"
      },
      {
         "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
         "size": 154,
         "digest": "sha256:00f7cedcc36e473a1df7df3f75b2e5963dc3dfe0c872372652d8e4eae072f307"
      },
      {
         "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
         "size": 10682058,
         "digest": "sha256:aad509e96da0e235a0a1166b95d0216b20c7e9055ccd8fa1855a0f1feb6c5faf"
      }
   ]
}
```

Here's what we get for a) computing the digest from the raw json string, b) computing the digest from our marshalled structured, c) the digest we downloaded during manifest selection (i.e. the digest told to us by the registry, our baseline for correctness)

![image](https://github.com/dotnet/sdk/assets/573979/a4c37076-d0e1-46d0-96de-fc17feb3fff6)

Reply:
After some hacking, it's _very_ hard to round-trip the manifest in System.Text.Json (or any potentially-lossy serializer, I suppose). We can't re-compute the manifest from our ManifestV2 structure and have any kind of fidelity, regardless of how many formatting tweaks I try to do. This also wouldn't be resilient - I'm pretty focused on mcr.microsoft.com for my testing but there doesn't seem to be any real spec for how the JSON should be formatted.

So to work around this, we're going to track the original digest when it's known and flow that through.

Reply:
[`c0fefe3` (#39160)](https://github.com/dotnet/sdk/pull/39160/commits/c0fefe3f5f07a699c074fb2219b806587af8f85d) should fix this - in it we 

* try to detect and set the digest for the manifest when we read it, and
* snapshot the base manifest in the ImageBuilder so that we can easily use the base manifests' digest instead of recomputing the digest of a mutable thing.